gcwatch.cpp

```diff
@@ -0,0 +1,205 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <atomic>
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <jvmti.h>
+
+static int port;
+
+static std::atomic_bool inGc;
+static std::atomic_long start;
+
+static long getMillis()
+{
+   struct timespec ts;
+   clock_gettime(CLOCK_MONOTONIC, &ts);
+   return (ts.tv_sec * 1000) + (ts.tv_nsec / (1000 * 1000));
+}
+
+static void JNICALL gcStart(jvmtiEnv *jvmti)
+{
+   start = getMillis();
+   inGc = true;
+}
+
+static void JNICALL gcFinish(jvmtiEnv *jvmti)
+{
+   inGc = false;
+}
+
+static int serverSocket()
+{
+   int fd = socket(AF_INET6, SOCK_STREAM, 0);
+   if (fd == -1) {
+      perror("ERROR: failed to create GcWatch socket");
+      exit(1);
+   }
+
+   int reuse = true;
+   if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse)) == -1) {
+      perror("ERROR: failed to set GcWatch socket option");
+      exit(1);
+   }
+
+   struct sockaddr_in6 addr = {};
+   addr.sin6_family = AF_INET6;
+   addr.sin6_port = htons(port);
+   addr.sin6_addr = in6addr_any;
+
+   if (bind(fd, (struct sockaddr *) &addr, sizeof(addr)) == -1) {
+      perror("ERROR: failed to bind GcWatch socket");
+      exit(1);
+   }
+
+   if (listen(fd, SOMAXCONN) == -1) {
+      perror("ERROR: failed to listen on GcWatch socket");
+      exit(1);
+   }
+   
+   return fd;
+}
+
+static void JNICALL worker(jvmtiEnv *jvmti, JNIEnv *jni, void *arg)
+{
+   int server = serverSocket();
+
+   fprintf(stderr, "GcWatch listener started on port %d\n", port);
+
+   while (true) {
+      int client = accept(server, NULL, 0);
+      if (client == -1) {
+         continue;
+      }
+
+      char output[128];
+      if (inGc) {
+         sprintf(output, "GC %ld\n", getMillis() - start);
+      }
+      else {
+         strcpy(output, "OK\n");
+      }
+
+      write(client, output, strlen(output));
+      close(client);
+   }
+}
+
+static jthread createThread(JNIEnv *jni, const char *threadName)
+{
+   jclass clazz = jni->FindClass("java/lang/Thread");
+   if (clazz == NULL) {
+      fprintf(stderr, "ERROR: GcWatch: failed to find Thread class\n");
+      exit(1);
+   }
+
+   jmethodID init = jni->GetMethodID(clazz, "<init>", "(Ljava/lang/String;)V");
+   if (init == NULL) {
+      fprintf(stderr, "ERROR: GcWatch: failed to find Thread constructor\n");
+      exit(1);
+   }
+
+   jstring name = jni->NewStringUTF(threadName);
+   if (name == NULL) {
+      fprintf(stderr, "ERROR: GcWatch: failed to create String object\n");
+      exit(1);
+   }
+
+   jthread object = jni->NewObject(clazz, init, name);
```

Comment:
do we need to set daemon for this thread?

Reply:
`RunAgentThread` creates daemon threads